rfcs/0180-broken-package-removal.md

```diff
@@ -0,0 +1,94 @@
+---
+feature: broken-package-removal
+start-date: 2024-07-14
+author: Jörg "Mic92" Thalheim
+co-authors:
+shepherd-team: @emilazy @preisi @jopejoe1
+shepherd-leader: @jopejoe1
+related-issues:
+---
+
+# Summary
+
+[summary]: #summary
+
+This RFC defines conditions under which we remove broken packages or unmaintained leaf packages.
+The RFC does not aim to define all possible reasons for removing a package but instead focuses on simple, automatable rules.
+Since this document serves as a guideline for automatic package removal, it also allows for the possibility of removing a package earlier for other reasons.
+
+# Motivation
+
+[motivation]: #motivation
+
+Broken and unmaintained packages still consume valuable time.
+Broken packages need to be evaluated to some extent, and unmaintained packages might still need to be rebuilt when one of their dependencies changes.
+Removing unmaintained or unused code should improve overall package quality and help save resources.
+Since adding packages is easy, removing them should also be simple, ideally automated.
+
+# Detailed Design
+
+[design]: #detailed-design
+
+All broken packages or unmaintained leaf packages will be removed after a full NixOS release cycle.
+For example, if `hello-unmaintained` is marked as unmaintained in 23.11, it can be removed from master after the 24.05 release.
+
+## Broken Packages
+
+Packages that are unconditionally marked as broken on all platforms are candidates for removal.
+If a package has dependent packages, those dependent packages will also be marked as indirectly broken.
+If the dependent packages cannot function without the broken package, they should also be removed.
+For example, if a library is broken and an application depends on it, the application will also be removed.
+
+## Unmaintained Packages
+
+We will add an "unmaintained" warning for all packages with an empty maintainer field that do not have any dependent packages.
```

Comment:
As I understand it, this "unmaintained" label + warning is what marks a package for removal in the next release but, as it's written, it sounds like any package that does not have maintainers listed is subject for removal in which case we wouldn't need such a label as the empty maintainers field is that label in effect.

This implies that there is a category of packages that have no maintainers but should not be removed which is unfortunately quite common in Nixpkgs. These are often important packages that are, in effect, communally maintained: @r-ryantm PRs are merged without maintainers and sometimes people like i.e. @trofi create PRs updating packages often quite close to the core that have no official maintainer.
   
This problem exacerbated by the recent rounds of maintainers leaving the project. If we had this policy before those events, we would need to remove a large amount of our core packages in 24.11 if we followed this RFC's rules. That's not feasible.

The RFC does not specify how this category of packages should be handled or recognise that it doesn't apply to that case and I think it must do either one of those.

Reply:
Note that the RFC only proposes the removal of unmaintained *leaf* packages, so most of these unmaintained core packages would never be considered for removal under these grounds.

Reply:
we are talking about automated PRs though, right? annoying if a bot keeps opening PRs to remove the same package and we close it

or is the intention to annoy people into maintaining said package? that might be effective... 🤔

Reply:
Automation is deferred to future work, but would presumably detect whether the packages are leaves before opening a PR.